### ISSUE TITLE
[expo-notifications] FCMv1 payload structure should be simplified

### ISSUE THREAD

**Issue by douglowder** [expo/expo#30533]
### Minimal reproducible example

https://github.com/brentvatne/microfoam-app

### What platform(s) does this occur on?

Android

### Where did you reproduce the issue?

in a standalone app

### Summary

Issue description from @dpuckett:

The payload mapping for received notifications is all over the place, and as far as I can tell doesn't conform to the standard set out by Firebase?

Firebase give multiple examples of new payload templates that should be used with the move to FCM v1 here:
https://firebase.google.com/docs/cloud-messaging/migrate-v1#update-the-payload-of-send-requests

The simplest of these being a platform agnostic payload with a message and some data:
```json
{
  "message": {
    "topic": "news",
    "notification": {
      "title": "Breaking News",
      "body": "New news story available."
    },
    "data": {
      "story_id": "story_12345"
    }
  }
}
```

Expos mapping of the new payload structure is currently mapping to totally differently named nodes, but also not following the expected format at all.
Please can this be addressed and the payload structure for received notifications using expo-notifications package be updated to follow Firebases examples?

It's an awful lot simpler than the payload structure we need to use to work around the current issues with expo-notifications, which looks like this:

```json
{
  "message": {
    "notification": {
      "title": "Breaking News",
      "body": "New news story available."
    },
    "data":{
        "message": "content message",
        "body":"{\"action\": \"$(action)\", \"itemid\": \"$(itemId)\"}"
    }
    
  }
}
```

### Environment

SDK 51.0.20 and later, expo-notifications 0.28.10 and later

### Expo Doctor Diagnostics

Not applicable

**Comment by muminbhat** [comment]
The data payload is included when a notification is sent from the Expo Push Server, however when I use Firebase Console directly to send notifications to both android and iOS the data payload is missing.

**Comment by nyenloop** [comment]
The same issue to me, data payload is missing, please help me !!!!!!

**Comment by prashantayal** [comment]
We tried to load the data from FCM payload but it is returning **null** in **content** object. However, if we are passing data in the **body** key, we are able to fetch the data in **dataString** key of the content object. But this is not the standard way of sending the payload as it should be the same across all the tech stacks (expo, react-native, flutter, ios, android).

However, iOS is working fine with no issues. Thanks

**Comment by krisgerhard** [comment]
Spent a whole day trying to make sense of this but here is our resume:

We are using Expo Push Api and receiving notifications on Android devices.

1. When sending push notification with `title`, `body`, `data`.
* `addNotificationReceivedListener` handles everything as expected. `notification.request.content.data` is exactly the JSON we sent. Eventhough remoteMessage has the data in string form and wrapped into `data.body`. We would like to only care about content.data so that's okay.
* `setNotificationHandler's handleNotification` puts our JSON to `notification.request.content.**dataString**` while `data` is not present at all. Notice that dataString is not part of the typescript model or documented anywhere.

2. When sending push notification with `data` but without `title`, `body` (silent notification).
* `addNotificationReceivedListener`  copies data from remoteMessage as a string and is wrapped into `data.body`, so getting the JSON we actually sent requires parsing JSON.
* `setNotificationHandler's handleNotification` now behaves exactly like `addNotificationReceivedListener` .

So in conclusion if you would like to parse JSON data from Notification object, then there really is no consistency.
A non-perfect workaround:
```
export const extractNotificationData = (
  expoNotification: Notifications.Notification,
) => {
  try {
    const content = expoNotification.request.content;
    if (content.hasOwnProperty('dataString')) {
      // @ts-ignore
      return JSON.parse(content.dataString);
    }
    if (content.hasOwnProperty('data')) {
      const contentData = content.data;
      if (contentData.hasOwnProperty('body')) {
        return JSON.parse(contentData.body);
      } else {
        return contentData;
      }
    }
    return {};
  } catch (e) {
    return {};
  }
};
```

**Comment by dpuckett** [comment]
@douglowder is there an update on this issue? We're 6 months down the road, and it seems this isn't being actively commented on.

**Comment by douglowder** [comment]
@dpuckett we will take a look. Over the past few months, there have been a lot of changes both in the `expo-notifications` package and in the Expo push service, to correctly support FCMv1, remove legacy code, and make sure that features like custom sounds, etc., are supported. 

Other considerations:
- The payload shape must support both iOS (APNS notifications) and Android (FCMv1 notifications)
- We have to be very careful not to make any change that breaks existing clients

@vonovak

**Comment by vonovak** [comment]
Hello @dpuckett and thanks for asking,

> The payload mapping for received notifications is all over the place, and as far as I can tell doesn't conform to the standard set out by Firebase?

For sending notifications via our [Push Service](https://docs.expo.dev/push-notifications/sending-notifications/), the payload specification is [here](https://docs.expo.dev/push-notifications/sending-notifications/#message-request-format).

For reading the data on the client, it's possible that there are multiple ways that the data comes out of on the client side, and we're always looking for ways to improve this. To do that we need comments that very specifically point out problems, such as [this one](https://github.com/expo/expo/issues/30533#issuecomment-2511519160). I'd like to politely point out that saying "mapping for received notifications is all over the place" doesn't provide information that would be actionable. I'd also like to highlight the fact that expo is open-source and you can be the driver behind a change you want to see.

> as far as I can tell doesn't conform to the standard set out by Firebase?

Firebase FCM has server-and client-side apis (I wouldn't call it a "standard" but I think I understand what you meant) for sending push notifications, and so do we - the Expo push service (link to payload structure shared above) on the server-side and the `expo-notifications` library on the client. It is not our goal to have APIs that are the same as Firebase's. 
We are not there yet, but our goal is having a transparent (not a black box), flexible API for sending push notifications across platforms, which supports most if not all features of FCM and APNs and is delightful to use.


> Expo's mapping of the new payload structure is currently mapping to totally differently named nodes, but also not following the expected format at all.

Kindly give detailed examples, but please note that it's not our goal to be a FCM clone.

Since notifications for Android are sent via FCM, we need to be compatible with FCM on the client side. In order to enable some cross-platform feature with unified apis on both platforms, we may inject our own content into the notification data, to enable the feature on the client-side. We develop the client-side package to be used in tandem with our server-side Push Service. If you choose to not use our Push Service but use FCM instead, that's completely fine, but there might be some cases where you'll need to read the client-side code to figure out how to use a feature with FCM backend. These apis are not documented but they are unlikely to change because we need to keep compatibility for old expo-notifications package versions.

Hope this helps to get some understanding!

Thank you 🙂

**Comment by dpuckett** [comment]
@vonovak Thank you for taking the time to reply.

I think you have misunderstood the actual problem being described in the op.

We are not using Expo services to send notifications. We are using Azure Notification Hubs to manage device registration data and send push notifications. Because of this, ANH specifies that the push payload sent follows the structure set out by Firebase for FCM V1 API.
That structure is what was referenced by @douglowder in his op on this thread.

The issue, is when using the Expo-Notifications library to receive these notifications, and the mapping from sent payload, to received payload. They do not align, and data is either lost or the notification not received at all. The data mapped from what should be a standard FCM V1 payload is being mapped to differently named properties within your libraries Notification type.

This becomes a problem because your documentation states that use of the Expo-Notifications library is compatible with other notification services, and this was the case, until your updates to the lib to incorporate the FCM V1 API.

Not mapping from the standardised payload structure given by Firebase will cause problems for multiple services trying to send to applications using your library to receive them on devices, and alienate people from using your open source libraries
(perhaps this is a strategic business goal, and if so that is your choice).

There was before a thread that had multiple examples of mismapping, I think in the original issue tracking thread?
If I recall, @douglowder has already done quite a lot of work to confirm which parts are not mapped correctly, which then led to him creating this ticket.
I am unable to find those notes at this time, but if @douglowder were able to link to them, they would be far more comprehensive that the small examples I can give.

**Comment by vonovak** [comment]
@dpuckett okay understood.

> Not mapping from the standardised payload structure given by Firebase will cause problems for multiple services trying to send to applications using your library to receive them on devices, and alienate people from using your open source libraries
(perhaps this is a strategic business goal, and if so that is your choice).

As I said above, we need to be compatible with FCM on the client side - that's a _necessity_ for us and doing otherwise would not be smart (it certainly isn't a strategic goal). If we're not compatible with FCM, that's a bug.

> This becomes a problem because your documentation states that use of the Expo-Notifications library is compatible with other notification services.

We claim that the client package is push-service agnostic because all push services, at the end of the day, are proxies over FCM and APNs. If sending a notification via our service works, it follows that sending an equivalent push notification over FCM and APNs will work too.

> The data mapped from what should be a standard FCM V1 payload is being mapped to differently named properties within your libraries Notification type.

To follow the same property naming as FCM is neither practical nor possible because of cross-platform support of our packages. We need unified naming that "works" for both Android and iOS.


If you give us examples of FCM messages that don't work with our client-side library, we will fix them because _we send our push messages via FCM and so we need it to work_. However, you're using another wrapper on top of FCM. If there's an issue with what Azure Notification Hub does with the payload, it's not in our ability to provide a workaround for that.

Hope this makes sense,
Vojtech

**Comment by fobos531** [comment]
> We are not there yet, but our goal is having a transparent (not a black box), flexible API for sending push notifications across platforms, which supports most if not all features of FCM and APNs and is delightful to use.

@vonovak While we're on this topic, I would like to bring up a thing that might be good to discuss internally within the Expo team - AFAIK right now the Expo push service is closed-source. It would be worth considering to open-source it as it would facilitate community contributions for features such as communication notifications on iOS, for example. See: https://developer.apple.com/documentation/usernotifications/implementing-communication-notifications

**Comment by vonovak** [comment]
@fobos531 thanks for asking; to tell the truth, the push notification service is too deeply integrated with our other services for us to be able to strip it out with a reasonable amount of effort.
 

For the feature that you mentioned though, you don't need anything extra afaict. The communication notifications are handled through a Notification Service Extension - that's a separate target in your app and it's something you can do already.

Thank you

**Comment by moneszarrugh** [comment]
@vonovak 

This is my workaround. Don't know why it isn't like this out of the box. I'm on SDK 51.

```ts
Notifications.addNotificationReceivedListener(
      (notification) => {
        const data = Platform.select({
          ios: notification.request.content.data,
          android: JSON.parse(notification.request.content.data.body),
       });
       // data is now consistant in ios and android
    }
);
```

**Comment by vonovak** [comment]
Hello and thanks for reporting,
the inconsistency with `dataString` being present instead of `data` was fixed in https://github.com/expo/expo/commit/e99fffa9310c002ac53aedcb5985ce38578b15da.

The fix applies to `addNotificationReceivedListener` and `setNotificationHandler`.

For [background tasks](https://docs.expo.dev/versions/latest/sdk/notifications/#registertaskasynctaskname), `dataString` is present on both platforms, and you need to JSON.parse it to obtain the object.
Thank you 🙂
